Autofill application fields

LeadFreeCandy opened the pull request:
# Pull Request

## Description

Please include a summary of the changes and the related issue (if applicable). Please also include relevant motivation and context. List any dependencies that are required for this change (if applicable).

Fixes # (issue)

## Type of change

- [ ] Bug fix (change which fixes an issue)
- [ ] Enhancement (change which alters existing functionality)
- [ ] New feature (change which adds new functionality)
- [ ] This change requires a wiki/documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

Try to checkoff as much as possible if not everything!

LeadFreeCandy committed:
Add langchain dependencies

LeadFreeCandy committed:
Use LangChain to automate autofill

LeadFreeCandy committed:
Make api requests in parallel

LeadFreeCandy committed:
switch to llm

timelord1102 committed:
further modified the autofill algorithm. Now MUCH more efficient runtime
and memory wise. Also added a new feature to autofill allowing for AI completion
of dropdown menus.

The algoritm continues to use GPT to generate the proper field response.
The new drop down algorithm passes GPT all possible options allowing it
to choose the appropriate response.

bug fix

LeadFreeCandy committed:
Autofill script runs an all frames

timelord1102 committed:
merges my autofill with Samir's

does not work as intented
```

LeadFreeCandy committed:
fix chain issue

timelord1102 committed:
modifies dropDown autofill to work again

modifies dropDowns autofill and answerField to be compatible with eachother
dropdown fields still needs a way to access dropdown dield info such as name and id

timelord1102 committed:
adds ability to autofill file fields in application

uses the file name to create a file object and then fills the respective
field with the file object

needs more testing. shows that a file has been added but unsure if the
file is actually being uploaded

anthonyfabius commented on autofill/autofillScript.ts:
```diff
@@ -1,8 +1,114 @@
-if (usernameField) (usernameField as HTMLInputElement).value = "yourUsername";
-if (passwordField) (passwordField as HTMLInputElement).value = "yourPassword";
+const user: UserInfo = {
+  firstName: "Samir",
+  lastName: "Beall",
+  email: "montea8@rpi.edu",
+  address: "1999 Burdett Ave, Troy, NY 12180",
+  phoneNumber: "(123) 456-78910",
+  zip: "12180",
+  null: "other",
```
Maybe switch this to placeholder data. Like this:
```javascript
const user: UserInfo = {
  firstName: "John",
  lastName: "Doe",
  email: "johndoe@rpi.edu",
  address: "123sesamestreet",
  phoneNumer: "(123) 456- 7890",
  zip: "12345",
  null: "other",
}
```

anthonyfabius commented on autofill/autofillScript.ts:
```diff
@@ -1,8 +1,114 @@
+  null: "other",
+};
+const files: FilePaths = {
+  resumePath: "C:\\Users\\lordo\\Downloads\\Ariels Resume-1.pdf",
+  transcriptPath: "C:\\Users\\lordo\\Downloads\\Academic Transcript.pdf",
+  coverLetterPath: "null",
```
Remove absolute pathing and use a placeholder name

anthonyfabius reviewed (changes requested):
Looks good! Just remove some personal data and replace it with placeholder stuff. 👍

LeadFreeCandy committed:
print saved options

LeadFreeCandy committed:
Merge remote-tracking branch 'origin/main' into Autofill-Application-Fields

LeadFreeCandy committed:
add additional options and integrate with autofill

timelord1102 committed:
Restructures autofill with a new data structure.

The new data structure makes the code slower but much mor accurate and easier to maintain.
One issue arises with the new llm type seach but i suspect the issue comes with the prompt.

timelord1102 committed:
Merge branch 'Autofill-Application-Fields' of https://github.com/EasyApp-RPI/EasyApp into Autofill-Application-Fields

timelord1102 committed:
Drastically optimizes the autofill script

Field search and fill are combined at every point possible, drastically
decreasing processing time

timelord1102 committed:
slightly changes the identification prompt

removing human prompt suggestions seems to have fixed an issue with
address being recognized as a file prompt. Everything else seems to continue
to work as expected.

timelord1102 committed:
removed file paths and random comment

timelord1102 committed:
comments out file filling until the data is ready

anthonyfabius commented on autofill/autofillScript.ts:
```diff
@@ -1,8 +1,233 @@
+  }
+}
+
+/*async function fileFields(data: inputElements) {
+  // get all inputElemets with a type of file
+  for (let i of data.inputs) {
+    let fieldInfo: FieldInfo = {
+      inputLabel: data.label.textContent || "",
+      name: data.inputs[0].name || "",
+      id: data.inputs[0].id || "",
+      placeholder: data.inputs[0].placeholder || "",
+      type: data.inputs[0].type || null,
+      header: data.header,
+    };
+   let file = await answerFile(files, fieldInfo);
+   if (file == "null") continue;
+
+   // get text after last '\\' in file path. This is the file name
+   let fileName = file.split("\\").pop();
+   // create a file object with the file name
+   const dataTransfer = new ClipboardEvent('').clipboardData || new DataTransfer();
+   dataTransfer.items.add(new File(["file"], fileName || ""));
+   // set input files to the file object, then clear the data transfer and continue to next field
+   i.files = dataTransfer.files;
+   dataTransfer.items.clear();
+  }
+}*/
```
💀

anthonyfabius reviewed (dismissed):
solid

timelord1102 committed:
Prettified Code!

Automated with github actions

timelord1102 committed:
Repairs dropdown/autofill allows for simulating user input and rudimentary checkbox support

Fixes an issue with dropdowns not setting correctly and autofill not working on random Fields
It also fixes the issue with some required fields needing user input to be set correctly
by simulating user input. Rudimentary checkbox support is also added, most likely buggy.

timelord1102 committed:
Merge branch 'Autofill-Application-Fields' of https://github.com/EasyApp-RPI/EasyApp into Autofill-Application-Fields

timelord1102 committed:
Prettified Code!

Automated with github actions

timelord1102 committed:
adds a delay before filling dropdowns

There is an issue with dropdowns that are dynamically populated.
The script will try to fill the dropdown before the options are available.
This adds a delay to allow the options to populate before filling the dropdown.

timelord1102 committed:
Merge branch 'Autofill-Application-Fields' of https://github.com/EasyApp-RPI/EasyApp into Autofill-Application-Fields

timelord1102 committed:
Prettified Code!

Automated with github actions

timelord1102 committed:
adds more types to basic autofill

timelord1102 committed:
Merge branch 'Autofill-Application-Fields' of https://github.com/EasyApp-RPI/EasyApp into Autofill-Application-Fields
